From c3ef1644b33786f398e5c2429273ea1ae8b8606e Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Thu, 25 Jun 2026 15:58:10 -0700
Subject: [PATCH 06/10] Fix uninitialized pointer dereference in
 MsftExtensionManager

A compromised Bluetooth controller could exploit an initialization race condition by sending a premature MSFT_LE_MONITOR_DEVICE_EVENT during stack bring-up. This triggers a virtual function call on an uninitialized `scanning_callbacks_` pointer, potentially leading to a control-flow hijack or Denial of Service in the privileged bluetooth process.

This change resolves the issue by explicitly default-initializing the `scanning_callbacks_` pointer to `nullptr` in `MsftExtensionManager::impl`. Additionally, a null check is added before invoking `OnTrackAdvFoundLost`. If the pointer is null, the event is safely dropped and a warning is logged.

Bug: 522372651
Test: built and atest locally
Flag: EXEMPT BUGFIX
This was autogenerated by MendIt (go/androidmendit)

Change-Id: I044a3cf62559a9e758c4d0f5d89525ae4574c57f
(cherry picked from commit d6505b430477940ff61355726ca3142fdd06fb21)
---
 system/gd/hci/msft.cc | 25 ++++++++++++++++---------
 1 file changed, 16 insertions(+), 9 deletions(-)

diff --git a/system/gd/hci/msft.cc b/system/gd/hci/msft.cc
index 4c48924f388..77c79af9d57 100644
--- a/system/gd/hci/msft.cc
+++ b/system/gd/hci/msft.cc
@@ -74,18 +74,25 @@ struct MsftExtensionManager::impl {
   void handle_le_monitor_device_event(MsftLeMonitorDeviceEventPayloadView view) {
     log::assert_that(view.IsValid(), "assert failed: view.IsValid()");
 
+    if (scanning_callbacks_ == nullptr) {
+      log::warn("scanning_callbacks_ is null, dropping Microsoft vendor event.");
+      return;
+    }
+
     // The monitor state is 0x00 when the controller stops monitoring the device.
-    if (view.GetMonitorState() == 0x00 || view.GetMonitorState() == 0x01) {
-      AdvertisingFilterOnFoundOnLostInfo on_found_on_lost_info;
-      on_found_on_lost_info.advertiser_address_type = view.GetAddressType();
-      on_found_on_lost_info.advertiser_address = view.GetBdAddr();
-      on_found_on_lost_info.advertiser_state = view.GetMonitorState();
-      on_found_on_lost_info.monitor_handle = view.GetMonitorHandle();
-      scanning_callbacks_->OnTrackAdvFoundLost(on_found_on_lost_info);
-    } else {
+    // The monitor state is 0x01 when the controller starts monitoring the device.
+    if (view.GetMonitorState() != 0x00 && view.GetMonitorState() != 0x01) {
       log::warn("The Microsoft vendor event monitor state is invalid.");
       return;
     }
+
+    AdvertisingFilterOnFoundOnLostInfo on_found_on_lost_info;
+    on_found_on_lost_info.advertiser_address_type = view.GetAddressType();
+    on_found_on_lost_info.advertiser_address = view.GetBdAddr();
+    on_found_on_lost_info.advertiser_state = view.GetMonitorState();
+    on_found_on_lost_info.monitor_handle = view.GetMonitorHandle();
+
+    scanning_callbacks_->OnTrackAdvFoundLost(on_found_on_lost_info);
   }
 
   void handle_msft_events(VendorSpecificEventView view) {
@@ -337,7 +344,7 @@ struct MsftExtensionManager::impl {
   hal::HciHal* hal_;
   hci::HciInterface* hci_layer_;
   Msft msft_;
-  ScanningCallback* scanning_callbacks_;
+  ScanningCallback* scanning_callbacks_{nullptr};
 };
 
 MsftExtensionManager::MsftExtensionManager(os::Handler* handler, hal::HciHal* hal,
-- 
2.55.0

